[Messenger] Add possibility to define routing key when sending message throught amqp

vincenttouzet opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.2
 Bug fix?      | yes
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | #28133
| License       | MIT

As describe in #28133 we can't actually define a custom routing key when sending message. It is now possible with the `Symfony\Component\Messenger\Stamp\AMQP\RoutingKeyStamp`

```php
$envelop = Envelop::wrap($message)->with(new RoutingKeyStamp('dummy_routing'));
```

vincenttouzet committed:
[Messenger] Add possibility to define routing key when sending message through amqp

brusin commented:
Guys, do you have a release date please? This PR is blocking for me and would be much appreciated :)

vincenttouzet commented:
@brusin I see that @nicolas-grekas added this to the next milestone. So it will probably be available in Symfony 4.2 that will be release in the end of november 2018

vincenttouzet commented:
I've updated the code to be compliant with the PR #28911

vincenttouzet commented on src/Symfony/Component/Messenger/Stamp/AmqpRoutingKeyStamp.php:
```diff
@@ -0,0 +1,30 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Stamp;
```
Like `Symfony\Component\Messenger\Stamp\AMQP` ?

sure i'll change this

sroze replied:
Within the transport: `Symfony\Component\Messenger\Transports\AmqpExt\Stamp` :)

vincenttouzet commented:
@sroze I've made the changes but I don't understand why the tests failed.

```
1) Symfony\Component\Messenger\Tests\Transport\AmqpExt\AmqpSenderTest::testItSendsWithRoutingKey
Error: Call to undefined method Symfony\Component\Messenger\Envelope::wrap()
```

This same code worked for other tests 🤔

sroze commented:
The `Envelope::wrap` method has been removed in `master` 🙃

sroze commented:
Actually, it should be re-introduced, it was a mistake, it's very convenient userland 🤔

vincenttouzet commented:
meh 😢 You merged the remove just after i rebased to master xD. I've re-rebased and updated my tests. Waiting for Travis.

I think the remove may add BC Break to existing application. (If I update my application from Symfony 4.1 to 4.2 it will break without the wrap method.)

nicolas-grekas commented:
FYI, I think this is not the approach that we should follo: adding this routing key is tightening the message to the transport, which defeats the purpose of the component. Instead, the message should be routed with a name, and the transport should be selected by name. I'm on a PR to prepare the neede code infrastructure, i'll keep you posted.

nicolas-grekas commented:
FYI, I think this is not the approach that we should follo: adding this routing key is tightening the message to the transport, which defeats the purpose of the component. Instead, the message should be routed with a name, and the transport should be selected by name. I'm on a PR to prepare the neede code infrastructure, i'll keep you posted.

nicolas-grekas commented:
OK, thinking twice about this, I'm withdrawing my previous comment. The routing can also be thought as a transport agnostic thing, eg something that specifies the user that some message should be sent to. Each transport can then use or not this info to route directly to some destination.

nicolas-grekas commented on src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpSender.php:
```diff
@@ -37,8 +38,13 @@ public function __construct(Connection $connection, SerializerInterface $seriali
     {
+        $routingKey = null;
+        /** @var RoutingKeyStamp|null $routingKeyConfig */
+        if ($routingKeyConfig = $envelope->get(RoutingKeyStamp::class)) {
```
$routingKeyStamp

nicolas-grekas commented on src/Symfony/Component/Messenger/Transport/AmqpExt/Stamp/RoutingKeyStamp.php:
```diff
@@ -0,0 +1,32 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\AmqpExt\Stamp;
```
should be moved to the root Stamp namespace: we should make this stamp transport agnostic (and it should be documented as such, doc PR welcome btw :) )

sroze replied:
For now, it's Amqp only so let's keep it here. Routing key makes little sense with Redis for example.

nicolas-grekas replied:
I don't agree with adding a transport specific stamp. That'd break the purpose of the component and the underlying patterns.
To me a routing key can be used eg to target a specific user. That's totally not amqp specific with the this pov.

nicolas-grekas reviewed (changes requested):
makes sense actually, here are some comments to generalize the concept

nicolas-grekas commented on src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php:
```diff
@@ -94,13 +94,13 @@ public static function fromDsn(string $dsn, array $options = array(), bool $debu
         }
 
-        $this->exchange()->publish($body, null, AMQP_NOPARAM, array('headers' => $headers));
+        $this->exchange()->publish($body, $routingKey, AMQP_NOPARAM, array('headers' => $headers));
```
I'd suggest enhancing the DSN that creates the connection object so that one could express/decide to ignore the routing key at "send time" and force routing without any routing key.
The purpose would be to force messages to go to a workers without taking care of the routing key - and *then* the worker would use the stamp to decide for the next hop destination.

vincenttouzet replied:
I'm not sure to understand. You prefer to leave the `routing_key` parameter on the publish method to null ?

Your idea is to have multiple transports configured ? For example if we want to handle log messages :

```yml
framework:
    messenger:
        transports:
            log.info: 
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%?routing_key=log.info'
                options:
                    exchange: {name: "log", type: "topic"}
                    queue: {name: "log.info", routing_key: "log.info"}
            log.error: 
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%?routing_key=log.error'
                options:
                    exchange: {name: "log", type: "topic"}
                    queue: {name: "log.error", routing_key: "log.error"}

        routing:
            # Route your messages to the transports
            'App\Message\LogMessage': [log.info, log.error]
```

A message would be sent to multiple transport and then each transport will decide if he need to handle the message or not ?

RabbitMQ already do this for us by routing the received message using the routing_key. We can then have only one transport for sending and one transport per queue for receiving.

```yml
framework:
    messenger:
        transports:
            # to send any log message
            log:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                options:
                    exchange: {name: "log", type: "topic"}
                    queue: {name: "log.info", routing_key: "log.*"}
            # Only for receiving using the messenger:consume-messages command
            log.info: 
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                options:
                    exchange: {name: "log", type: "topic"}
                    queue: {name: "log.info", routing_key: "log.info"}
            log.error: 
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                options:
                    exchange: {name: "log", type: "topic"}
                    queue: {name: "log.error", routing_key: "log.error"}

        routing:
            # Route your messages to the transports
            'App\Message\LogMessage': [log]
```

And the developer just need to create the message with its routing_key

```php
// The routing key is computed regarding the severity of the log message
$bus->dispatch(new Envelop(new \App\Message\LogMessage(...), new RoutingKeyStamp('log.error')));
```

I understand that it is thightly coupled with AMQP and therefor this is not agnostic :thinking:

nicolas-grekas commented:
Radically simpler approach. Less flexible but maybe only what you need for now: can you check if #28955 would to it for you?

vincenttouzet commented:
@nicolas-grekas I think this is the right way to achieve this.

With your modification we can then have : 

```
framework:
    messenger:
        transports:
            log.info: 
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                options:
                    exchange: {name: "log", type: "topic"}
                    queue: {name: "log.info", routing_key: "log.info"}
            log.error: 
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                options:
                    exchange: {name: "log", type: "topic"}
                    queue: {name: "log.error", routing_key: "log.error"}

        routing:
            # Route your messages to the transports
            'App\Message\LogMessage': [log.info, log.error]
```

But I wonder how can we check in the handler side how to handle only if it's from a given queue. For example the handler of the `App\Message\LogMessage` may be different if the message comes from the `log.error` queue or the `log.info`. Maybe with a `Symfony\Component\Messenger\Handler\MessageSubscriberInterface` but we can only define method and priority

nicolas-grekas commented:
Shall we close this PR then?
#28955 fixes the issue that was spotted right?
I think my comments above go in a different direction, which could be interesting to follow IMHO, but maybe later on, WDYT?

vincenttouzet commented:
I think it could be a good idea to explore the way you describe :+1: maybe we can discuss about it at forum PHP today ?

sroze commented:
Sure :)

On Fri, 26 Oct 2018 at 07:32, Vincent Touzet <notifications@github.com>
wrote:

> I think it could be a good idea to explore the way you describe 👍 maybe
> we can discuss about it at forum PHP today ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/symfony/symfony/pull/28772#issuecomment-433291352>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAxHEfJj4747RjUrAAojLG483_enVKSDks5uop5tgaJpZM4XNb8->
> .
>
